spark2/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,309 @@
+
+    LOG.info("Processing data from {} to {}", startOffset, endOffset);
+    List<FileScanTask> tasks = pendingBatches.stream()
+        .flatMap(batch -> batch.tasks().stream())
+        .collect(Collectors.toList());
+    CloseableIterable<FileScanTask> splitTasks = TableScanUtil.splitFiles(CloseableIterable.withNoopClose(tasks),
+        splitSize());
+    return Lists.newArrayList(
+        TableScanUtil.planTasks(splitTasks, splitSize, splitLookback, splitOpenFileCost));
+  }
+
+  private StreamingOffset calculateStartingOffset() {
+    StreamingOffset startingOffset;
+    if (startSnapshotId != null) {
+      startingOffset = new StreamingOffset(startSnapshotId, 0, true, false);
+    } else {
+      List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+      if (snapshotIds.isEmpty()) {
+        // there's no snapshot currently.
+        startingOffset = StreamingOffset.START_OFFSET;
+      } else {
+        startingOffset = new StreamingOffset(snapshotIds.get(snapshotIds.size() - 1), 0, true, false);
+      }
+    }
+
+    return startingOffset;
+  }
+
+  private StreamingOffset calculateEndOffset(StreamingOffset start) {
+    if (start.equals(StreamingOffset.START_OFFSET)) {
+      return StreamingOffset.START_OFFSET;
+    }
+
+    // Spark will invoke setOffsetRange more than once. If this is already calulated, use the cached one to avoid
+    // calculating again.
+    if (cachedPendingBatches == null || !cachedPendingBatches.first().equals(start)) {
+      this.cachedPendingBatches = Pair.of(start, getChangesWithRateLimit(start, maxSizePerBatch));
+    }
+
+    List<MicroBatch> batches = cachedPendingBatches.second();
+    MicroBatch lastBatch = batches.isEmpty() ? null : batches.get(batches.size() - 1);
+
+    if (lastBatch == null) {
+      return start;
+    } else {
+      boolean isStarting = lastBatch.snapshotId() == start.snapshotId() && start.isStartingSnapshotId();
+      return new StreamingOffset(lastBatch.snapshotId(), lastBatch.endFileIndex(), isStarting,
+          lastBatch.lastIndexOfSnapshot());
+    }
+  }
+
+  @VisibleForTesting
+  @SuppressWarnings("checkstyle:HiddenField")
+  List<MicroBatch> getChangesWithRateLimit(StreamingOffset startOffset, long maxSize) {
+    List<MicroBatch> batches = Lists.newArrayList();
+    long currentLeftSize = maxSize;
+    MicroBatch lastBatch = null;
+
+    checkOverwrite(table.snapshot(startOffset.snapshotId()));
+    if (shouldGenerateFromStartOffset(startOffset)) {
+      currentLeftSize -= generateMicroBatch(startOffset.snapshotId(), startOffset.index(),
+          startOffset.isStartingSnapshotId(), currentLeftSize, batches);
+      lastBatch = batches.get(batches.size() - 1);
+    }
+
+    // Current snapshot can already satisfy the size needs.
+    if (currentLeftSize <= 0L || (lastBatch != null && !lastBatch.lastIndexOfSnapshot())) {
+      return batches;
+    }
+
+    long currentSnapshotId = table.currentSnapshot().snapshotId();
```

Comment:
Do we need refresh table?

Reply:
I think we already did it in `setOffsetRange`.

Reply:
Thanks, I see, it is called in every micro batch.

Reply:
Yeah, it has to be refreshed every batch to make the table consuming moving forward.